### PR TITLE
Optimizing list formatter data deduplication

### PR DESCRIPTION
**Pull request by robertbastian**
This replaces the rigid fallback structure by a data-defined structure which is optimized for each language.

### CODE REVIEW

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (Manishearth) [inline experimental/list_formatter/src/deduplicating_array.rs:19]: It is technically possible for the custom derive to handle this case: when I implemented generic handling I explicitly punted this for later, but it's not too tricky to make happen.

I might need to add an attribute like:

```rust
#[derive(Yokeable, ZCF)]
enum Dedupe<#[yoke(yokeable_param) D> {..}
```

thoughts, @sffc ? I can add this if folks want.

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (sffc) [inline experimental/list_formatter/src/deduplicating_array.rs:1]: Is it expected that the postcard file is +3.69 KB with this change?  We should be able to make it a no-op, right?

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (sffc) [inline experimental/list_formatter/src/deduplicating_array.rs:20]: Thought: I don't know for sure how Postcard handles this, but when `T` is large, it means that an array of `[T; 12]` has to allocate space for 12 `T`s even if all but one of them is null.  Hopefully Postcard optimizes this, but I don't know. I suspect that there is some amount of indirection due to the use of an enum, which may explain the postcard size increase.

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (sffc) [inline experimental/list_formatter/src/deduplicating_array.rs:11]: Praise: I think this is a nice useful generic class to introduce.  We may want to move it somewhere more central after this is all checked in and landed.

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (sffc) [inline experimental/list_formatter/src/deduplicating_array.rs:16]: Thought: Another way to achieve the deduplication goal would be to handle it all at the Serde level.

Keeping in mind that Postcard is not self-describing, you could write out the following Postcard constructions to deduplicate a list of `N` elements:

1. Number of unique elements `M`
2. Sequence of `N` elements from 0 to `M`
3. Sequence of `M` elements containing the actual data

When deserializing, use the following algorithm to make it zero-copy:

1. Read `M`
2. Read the sequence of indexes into a stack array of fixed length
3. Create an uninitialized stack array of elements
4. When reading each element from Serde, set it into each element of the list having that index according to the sequence of indexes you stored on the stack in step 2

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (sffc) [reply experimental/list_formatter/src/deduplicating_array.rs:19]: Seems useful, but let's align on the architecture of the class first.

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (robertbastian) [reply experimental/list_formatter/src/deduplicating_array.rs:1]: I don't understand where this is coming from. One thing I removed was the default fallback, but the ", " ListJoinerPattern should only be 4 bytes (1 string length, 2 string, 1 index), and because of deduping that should only appear once per language. That said, I'm removing the discriminants of the Optionals, so that should be a 12 byte decrease per language.

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (robertbastian) [reply experimental/list_formatter/src/deduplicating_array.rs:11]: I think until we can make this generic in the length this isn't really reusable. Unless we make this whole file a macro.

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (robertbastian) [reply experimental/list_formatter/src/deduplicating_array.rs:16]: That sounds like a good solution. Currently I'm using untagged enums, so Serde tries to deserialize the variants from top to bottom and falls through if one fails. This means DeduplicatingArray<numeric> will probably not work, because it would deserialize the fallbacks as values (but that type doesn't make much sense anyway).

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (robertbastian) [reply experimental/list_formatter/src/deduplicating_array.rs:20]: Does `T` have a fixed size in postcard? It contains strings, are those offsets into a data section, or are the bytes encoded inline?
`Dedupe<T>` should have postcard-length 1 for the `Fallback` case, and the same length as `T` for the `Value` case due to the `untagged` setting.

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (robertbastian) [reply experimental/list_formatter/src/deduplicating_array.rs:19]: I don't think this type needs to be `Yokeable` though? It's an implementation detail, and `DeduplicatingArray` can be yoked and zcf'ed without it.

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (robertbastian) [reply experimental/list_formatter/src/deduplicating_array.rs:16]: Also, any overhead we add will be relatively expensive, because most patterns are < 10 bytes.

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (Manishearth) [reply experimental/list_formatter/src/deduplicating_array.rs:16]: Well, if this was Yokeable then DeduplicatingArray could use `#[derive(Yokeable)]` instead of having to do the manual thing

**Comment by robertbastian** [comment]
.

**Review by Manishearth — COMMENTED** [review]
not finished with the review but i figured i'd drop the comments i have so far

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (robertbastian) [reply experimental/list_formatter/src/deduplicating_array.rs:77]: They are slightly different though. This one holds `&T` whereas the one used for deserializing holds `T`.

Comment (Manishearth) [reply experimental/list_formatter/src/deduplicating_array.rs:77]: Ah, I see. Still, there's a struct and an impl, perhaps they should be named differently but outside.

Comment (robertbastian) [reply experimental/list_formatter/src/deduplicating_array.rs]: Done

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (robertbastian) [reply experimental/list_formatter/src/deduplicating_array.rs]: I'm using the array [index] for fallbacks now. That means arrays of arrays of single numbers won't be deduplicable, but that's not a big loss.

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (sffc) [inline experimental/list_formatter/src/deduplicating_array.rs:48]: Nit: Return an error instead of panicking, because we don't want malicious data to be able to crash an app.

**Review by Manishearth — DISMISSED** [review]
LGTM, though I'd also wait for shane

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (sffc) [inline experimental/list_formatter/src/deduplicating_array.rs:16]: Nit: `j` is an element in the closure, not an index.

```suggestion
        match array.iter().take(i).position(|value| &array[i] == value) {
```

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (sffc) [inline experimental/list_formatter/src/deduplicating_array.rs:114]: Nit/Suggestion: I think `HumanSer` / `HumanDe` are probably overkill, and you should use `MachineSer` / `MachineDe` for both cases.  The code would be more straightforward if you got rid of `HumanSer` / `HumanDe`, and it's not clear what you really gain from this separation.  If you let JSON do the externally tagged representation, as it should do with the machine versions, you would get something like

```json
[
  {
    "Value": {
      "default": {
        "string": ", ",
        "index_1": 2
      },
      "special_case": null
    }
  },
  {
    "Fallback": 0
  },
  {
    "Value": {
      "default": {
        "string": ", o ",
        "index_1": 4
      },
      "special_case": null
    }
  },
  {
    "Value": {
      "default": {
        "string": " o ",
        "index_1": 3
      },
      "special_case": null
    }
  },
  {
    "Fallback": 0
  }
]
```

I think this is not much worse than what you currently have:

```json
[
  {
    "default": {
      "string": ", ",
      "index_1": 2
    },
    "special_case": null
  },
  [
    0
  ],
  {
    "default": {
      "string": ", o ",
      "index_1": 4
    },
    "special_case": null
  },
  {
    "default": {
      "string": " o ",
      "index_1": 3
    },
    "special_case": null
  },
  [
    0
  ]
]
```

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (sffc) [inline experimental/list_formatter/src/deduplicating_array.rs:136]: Note: You should fill in these strings in order to enable JSON serialization of MachineSer to work.

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (sffc) [inline experimental/list_formatter/src/deduplicating_array.rs:176]: Note: If you accept my suggestion above, I *think* this code would need to change to expect a `u32` as the variant for machine-readable, and a `String` as the variant for human-readable.

**Review by sffc — COMMENTED** [review]
Praise: This is a very tight and clean implementation!  I think this is landable as-is, but I left a suggestion for how to improve it further.

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (robertbastian) [reply experimental/list_formatter/src/deduplicating_array.rs:114]: The bincode encoding actually maps to this JSON
```json
[
  {
    "Value": {
      "default": {
        "string": ", ",
        "index_1": 2
      },
      "special_case": null
    }
  },
  {
    "Fallback: 0": {}
  },
  {
    "Value": {
      "default": {
        "string": ", o ",
        "index_1": 4
      },
      "special_case": null
    }
  },
  {
    "Value": {
      "default": {
        "string": " o ",
        "index_1": 3
      },
      "special_case": null
    }
  },
  {
    "Fallback: 0": {}
  }
]
```
i.e. the fallback is encoded in the variant (this lets us always save a byte, and support `usize` indices with varint encoding). I think this looks pretty bad, but in any case it's not technically possible, as the variant string has to be `&'static str`. And if we need a separate formatter for JSON anyway, we might as well make it look nice.

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (robertbastian) [reply experimental/list_formatter/src/deduplicating_array.rs:136]: As explained above, it's not possible to fill in the variant string.

File: experimental/list_formatter/src/deduplicating_array.rs
Comment (sffc) [reply experimental/list_formatter/src/deduplicating_array.rs:114]: Nit: I still prefer a tagged JSON representation.

**Review by sffc — DISMISSED** [review]

**Review by sffc — DISMISSED** [review]

**Comment by sffc** [comment]
Please fix the features CI:

```
    Building crate=icu_list features=[]
   Compiling formatted_string_builder v0.1.0 (/home/runner/work/icu4x/icu4x/experimental/formatted_string_builder)
   Compiling zerovec v0.5.0 (/home/runner/work/icu4x/icu4x/utils/zerovec)
   Compiling icu_list v0.1.0 (/home/runner/work/icu4x/icu4x/experimental/list_formatter)
error[E0433]: failed to resolve: use of undeclared crate or module `serde`
 --> experimental/list_formatter/src/deduplicating_array.rs:1:5
  |
1 | use serde::de::{Deserialize, Deserializer, EnumAccess, Visitor};
  |     ^^^^^ use of undeclared crate or module `serde`
```

To reproduce, build `icu_list` with `--no-default-features`

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
This needs a data regen, but I think it's mergeable after that.

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
I believe shane's review was also green before the push (I really wish GitHub would retain old review state), so hitting the merge button.

**Comment by Manishearth** [comment]
Thanks!

**Review by sffc — COMMENTED** [review]
LGTM except I still prefer a tagged representation instead of the array thing

**Pull request merged**
